### PR TITLE
Fixed issue BTS-624

### PR DESCRIPTION
**Pull request by cpjulia**
### Scope & Purpose

Enterprise companion PR: https://github.com/arangodb/enterprise/pull/809

Bug-fix related to the ticket https://arangodb.atlassian.net/browse/BTS-624.
The `move-calculations-up` optimization rule is now also applied to subqueries, when they don't have dependencies on the outer nodes, don't have modification nodes and don't read their own writes. This fixed the execution of a query without the splicing-subqueries option being faster than the execution of a query with this option (after version 3.8, this option cannot be switched off). 

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] Backport for 3.9: https://github.com/arangodb/arangodb/pull/15238
- [x] Backport for 3.8: https://github.com/arangodb/arangodb/pull/15013
- [x] Backport for 3.7: https://github.com/arangodb/arangodb/pull/15019

#### Related Information

- [x] Docs PR: https://github.com/arangodb/docs/pull/795
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/809
- [x] GitHub issue / Jira ticket number: https://arangodb.atlassian.net/browse/BTS-624
- [ ] Design document: 

### Testing & Verification

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [x] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

File: arangod/Aql/OptimizerRules.cpp [hunk 1]
```diff
@@ -1780,7 +1795,7 @@ void arangodb::aql::moveCalculationsUpRule(Optimizer* opt,
         // transfer the calculation result to the coordinator instead of the
         // full documents
 
-        if (!::accessesCollectionVariable(plan.get(), nn, vars)) {
+        if (!isAccessCollection) {
           // not accessing any collection data
           break;
         }
```
[the remark this answers is not in the document]
Comment (cpjulia) [reply arangod/Aql/OptimizerRules.cpp:1801]: Agree

File: tests/js/server/aql/aql-optimizer-rule-move-calculations-up.js [hunk 2]
```diff
@@ -126,6 +143,134 @@ function optimizerRuleTestSuite () {
...
+      assertEqual(nodesDisabled.map(node => node.type), ["EnumerateListNode", "SubqueryStartNode", "EnumerateListNode",
+        "SubqueryStartNode", "EnumerateCollectionNode", "LimitNode", "SubqueryEndNode", "UpsertNode",
+        "SubqueryEndNode", "ReturnNode"]);
+      assertEqual(nodesDisabled[0].outVariable.name, "i");
+      assertEqual(nodesDisabled[1].subqueryOutVariable.name, "x");
+      assertEqual(nodesDisabled[2].outVariable.name, "j");
+      assertEqual(nodesDisabled[8].outVariable.name, "x");
+      assertEqual(nodesDisabled[9].inVariable.name, "x");
+
+    },
```
[the remark this answers is not in the document]
Comment (cpjulia) [reply tests/js/server/aql/aql-optimizer-rule-move-calculations-up.js:272]: Addressed

File: tests/js/server/aql/aql-optimizer-rule-move-calculations-up.js [hunk 3]
```diff
@@ -42,9 +44,24 @@ function optimizerRuleTestSuite () {
   var paramNone   = { optimizer: { rules: [ "-all" ] } };
   var paramEnabled    = { optimizer: { rules: [ "-all", "+" + ruleName ] } };
   var paramDisabled  = { optimizer: { rules: [ "+all", "-" + ruleName ] } };
+  var noMoveCalculationsUp = { optimizer: { rules: [ "+all", "-" + ruleName, "-" + ruleName + "-2"] } };
```
[the remark this answers is not in the document]
Comment (cpjulia) [reply tests/js/server/aql/aql-optimizer-rule-move-calculations-up.js]: Addressed

File: tests/js/server/aql/aql-optimizer-rule-late-document-materialization-arangosearch-primary-sort.inc [hunk 4]
```diff
@@ -176,8 +176,8 @@
                     "LET e = SUM(FOR c IN " + vn + " LET p = CONCAT(c.foo, c.foo) RETURN p) " +
                     "SORT CONCAT(a, e) LIMIT 10 RETURN d";
         let plan = AQL_EXPLAIN(query).plan;
+        assertNotEqual(-1, plan.rules.indexOf(ruleName));
         if (!isCluster) {
```
Comment (Dronplane) [inline tests/js/server/aql/aql-optimizer-rule-late-document-materialization-arangosearch-primary-sort.inc:180]: Should we then remove "isCluster" condition and indeed check late materialization rule usage in cluster in this case ?

File: tests/js/server/aql/aql-optimizer-rule-late-document-materialization.js [hunk 5]
```diff
@@ -342,8 +342,8 @@ function lateDocumentMaterializationRuleTestSuite () {
                     "LET e = SUM(FOR c IN " + collectionNames[(i + 1) % numOfCollectionIndexes] + " LET p = CONCAT(c.obj.b, c.obj.a) RETURN p) " +
                     "SORT CONCAT(a, e) LIMIT 10 RETURN d";
         let plan = AQL_EXPLAIN(query).plan;
+        assertNotEqual(-1, plan.rules.indexOf(ruleName));
         if (!isCluster) {
```
Comment (Dronplane) [inline tests/js/server/aql/aql-optimizer-rule-late-document-materialization.js:346]: Should we enable full check for the query result in the cluster case? As the rule is now applied and we have to validate the results

**Review by Dronplane — APPROVED** [review]
LGTM

**Pull request merged**